### PR TITLE
Fix some issues with comparison op codegen

### LINKED ISSUE #733: != operator not working with Integer Comparison
Hello
I have a simple stream query as follows

```sql
SELECT after_hash, before_hash, CAST(after_hash AS varchar) != CAST(before_hash AS varchar) FROM tirsdms_stream_test_2
```

Without casting the two columns (after_hash and before_hash) to String the `!=` operator returns true in all cases. Comparing two integer columns with `!=` does not work as expected.

### PR DESCRIPTION
**Pull request by rodesai**
Use the right operations for the various operand types:
    - For non-boolean scalars, use (<= && >=) to check for equality.
      == does a reference check rather than a value check. .equals
      requires the argument to be of the same type as the object, so
      equality comparisons between different types dont work. (e.g.
      BIGINT and INT).
    - For boolean scalars, only support == and !=. Implement using
      Boolean.compare
    - For Strings, use .equals for == and !=, and compareTo for other
      operators.
    - For array/map, only support == and != and implement using
      .equals

This patch also includes handling of null operands. The implemented
behaviour is informed by Postgres semantics for null operands. For
every operator other than IS DISTINCT FROM, return false if either
operand is null. Postgres actually returns null, but janino refuses to
accept null for a Boolean, so we return false for now. For
IS DISTINCT FROM, return true if exactly one of the operators is null,
false if they are both null, and defer to != otherwise.

Fixes #733

### CODE REVIEW

File: ksql-engine/src/test/java/io/confluent/ksql/codegen/CodeGenRunnerTest.java [hunk 1]
```diff
@@ -63,9 +109,228 @@ private Analysis analyzeQuery(String queryStr) {
         return analysis;
     }
 
+    public boolean evalBooleanExpr(String queryFormat, int cola, int colb, Object values[])
```
Comment (dguy) [inline ksql-engine/src/test/java/io/confluent/ksql/codegen/CodeGenRunnerTest.java]: nit: `private`?

File: ksql-engine/src/test/java/io/confluent/ksql/codegen/CodeGenRunnerTest.java [hunk 2]
```diff
@@ -63,9 +109,228 @@ private Analysis analyzeQuery(String queryStr) {
...
+        Assert.assertTrue(idx1 == cola || idx1 == colb);
+        Assert.assertNotEquals(idx0, idx1);
+        if (idx0 == colb) {
+            Object tmp = values[0];
+            values[0] = values[1];
+            values[1] = tmp;
+        }
+        Assert.assertEquals(expressionEvaluatorMetadata0.getUdfs().length, 2);
+        Object result0 = expressionEvaluatorMetadata0.getExpressionEvaluator().evaluate(values);
+        Assert.assertTrue(result0 instanceof Boolean);
```
Comment (dguy) [inline ksql-engine/src/test/java/io/confluent/ksql/codegen/CodeGenRunnerTest.java]: `assertThat(result0, instanceOf(Boolean.class))`

File: ksql-engine/src/test/java/io/confluent/ksql/codegen/CodeGenRunnerTest.java [hunk 3]
```diff
@@ -63,9 +109,228 @@ private Analysis analyzeQuery(String queryStr) {
...
+            throws Exception {
+        String simpleQuery = String.format(queryFormat, cola, colb);
+        Analysis analysis = analyzeQuery(simpleQuery);
+
+        ExpressionMetadata expressionEvaluatorMetadata0 = codeGenRunner.buildCodeGenFromParseTree
+            (analysis.getSelectExpressions().get(0));
+        Assert.assertTrue(expressionEvaluatorMetadata0.getIndexes().length == 2);
+        int idx0 = expressionEvaluatorMetadata0.getIndexes()[0];
+        int idx1 = expressionEvaluatorMetadata0.getIndexes()[1];
+        Assert.assertTrue(idx0 == cola || idx0 == colb);
```
Comment (dguy) [inline ksql-engine/src/test/java/io/confluent/ksql/codegen/CodeGenRunnerTest.java]: `assertThat(idx0, anyOf(equalTo(cola), equalTo(colb)))` ?
similarly on next line

File: ksql-engine/src/test/java/io/confluent/ksql/codegen/CodeGenRunnerTest.java [hunk 4]
```diff
@@ -63,9 +109,228 @@ private Analysis analyzeQuery(String queryStr) {
...
+            values[0] = values[1];
+            values[1] = tmp;
+        }
+        Assert.assertEquals(expressionEvaluatorMetadata0.getUdfs().length, 2);
+        Object result0 = expressionEvaluatorMetadata0.getExpressionEvaluator().evaluate(values);
+        Assert.assertTrue(result0 instanceof Boolean);
+        return (Boolean)result0;
+    }
+
+    public boolean evalBooleanExprEq(int cola, int colb, Object values[]) throws Exception {
```
Comment (dguy) [inline ksql-engine/src/test/java/io/confluent/ksql/codegen/CodeGenRunnerTest.java]: `private`?

Same for other methods in this class

**Comment by rodesai** [comment]
I thought about testing from SqlToJavaVistitorTest instead, but went with the CodeGenRunner tests instead since they actually verify that the resulting behavior is correct.

**Pull request merged**